### DISCUSSION TITLE
Maintaining a Plugins Repository For Headless Mode

### DISCUSSION THREAD

**Discussion by alexander-hanel** [Vector35/binaryninja-api#3473] [API] [answered]
Hello, 

Binary Ninja Version 3.1.3469 (Build ID 93650e26)

I'm looking to maintain a repository of plugins for use in headless mode. I'm trying to add my own local plugin repository using the API Repository Manager. Below I'm using the `official` plugins directory as an example POC. It appears none of the plugins are populated and it also references the repository by the string of the file path passed.  

```
% pwd 
/removed/Library/Application Support/Binary Ninja/repositories/official
% ls 
plugins		plugins.json
% python3.8
>>> from binaryninja import RepositoryManager
>>> mgr = RepositoryManager()
>>> status = mgr.add_repository("xyz", repopath="plugins.json")
>>> status
True
>>> mgr.repositories
[<community>, <official>, <plugins.json>]
>>> mgr.repositories[2].plugins
[]
```
This issue looks similar to [issue 2987](https://github.com/Vector35/binaryninja-api/issues/2987)?  I'm a little new to Binary Ninja's API so I was wondering if this was an error on my part or is this similar to issue 2987. I tried the recommendations in the issue but it didn't resolve my issue. Is using  `mgr.add_repository` for adding and running plugins the right approach or is there another method I should attempt? 

Thanks, 

Alex

**Comment by plafosse** [comment, marked as answer]
The first parameter passed to `add_repository` should be a URL which will download a `plugins.json` the second parameter is the directory name in the `Binary Ninja/repositories/` directory.

https://api.binary.ninja/binaryninja.pluginmanager-module.html#binaryninja.pluginmanager.RepositoryManager.add_repository

**Reply by alexander-hanel** [reply]
Thank you, this is a helpful response. I was under the false assumption that `repopath` was for a local file path. Is it possible to add a plugins to Binary Ninja using a file path rather than a URL repository path?  I believe I can copy the directory to the plugin directory but I was attempting to do everything via the API.

**Reply by psifertex** [reply]
It's not possible to add using a file path -- copying to the plugin directory would be the way to do that if you don't want to use the repository API.

**Comment by plafosse** [comment]
Additionally I'd recommend using the "unoffical" plugins settings. `pluginManager.unofficialName`

**Comment by alexander-hanel** [comment]
Thanks, I looked into `pluginManager.unofficialName` but I couldn't find any examples of its usage and as I mentioned i'm still new to the binja's APIs.
